pydantic/_internal/_typing_extra.py

```diff
@@ -192,43 +222,68 @@ def parent_frame_namespace(*, parent_depth: int = 2, force: bool = False) -> dic
     # note, we don't copy frame.f_locals here (or during the last return call), because we don't expect the namespace to be modified down the line
     # if this becomes a problem, we could implement some sort of frozen mapping structure to enforce this
     if force:
-        return frame.f_locals
+        return ImmutableNs(frame.f_locals)
 
     # if either of the following conditions are true, the class is defined at the top module level
     # to better understand why we need both of these checks, see
     # https://github.com/pydantic/pydantic/pull/10113#discussion_r1714981531
     if frame.f_back is None or frame.f_code.co_name == '<module>':
         return None
 
-    return frame.f_locals
+    return ImmutableNs(frame.f_locals)
 
 
-def get_module_ns_of(obj: Any) -> dict[str, Any]:
-    """Get the namespace of the module where the object is defined.
+class NsResolver:
+    """Protects namespaces from modifications and provides a way to merge namespaces without copying namespaces."""
 
-    Caution: this function does not return a copy of the module namespace, so it should not be mutated.
-    The burden of enforcing this is on the caller.
-    """
-    module_name = getattr(obj, '__module__', None)
-    if module_name:
+    def __init__(self, *namespaces: NsResolver | ImmutableNs | None) -> None:
+        self._namespaces: list[ImmutableNs] = []
+        self._resolved_memo: ImmutableNs | None = None
+
+        for ns in namespaces:
+            if ns is not None:
+                if isinstance(ns, NsResolver):
+                    self._namespaces.extend(ns._namespaces)
+                else:
+                    assert isinstance(ns, ImmutableNs)
+                    self._namespaces.append(ns)
+
+    @classmethod
+    def from_module_ns_of(cls, obj: Any) -> Self:
+        """Get immutable namespace of the module where the object is defined."""
+        return cls(ImmutableNs.from_module_ns_of(obj))
+
+    @contextmanager
+    def push(self, for_type: type[Any]):
+        """Push namespace of for_type as lowest priority in the resolved namespaces"""
```

Comment:
See my comment above - I think `push` should make the new take the highest priority so that this is intuitively like a revertable `dict` update

Reply:
The logic actually requires it to be "lowest priority". Some tests fail if it's other way around. Should we rename this into something like "prepend_ns_from(type)"?

Reply:
Which tests fail? I recall in the types_ns_stack context management that we give priority to namespaces in the way that you've replicated, but I'm not sure why we did that (and I don't think we should). Seems much more intuitive for later ns entries to take priority. Wondering what the use case is here.

Reply:
These fail if we instead do `self._namespaces.append(type_ns)` + `self._namespaces.pop()` (which is different from original as the priority is other way around):
```
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┓
┃  File                           ┃  Function                                                             ┃  Function Line  ┃  Error Line  ┃  Error            ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━┩
│  tests/test_dataclasses.py      │  test_base_dataclasses_annotations_resolving_with_override[pydantic]  │  1675           │  1714        │  ValidationError  │
│  tests/test_dataclasses.py      │  test_base_dataclasses_annotations_resolving_with_override[stdlib]    │  1675           │  1714        │  ValidationError  │
│  tests/test_main.py             │  test_extra_validator_named                                           │  2943           │  2954        │  ValidationError  │
│  tests/test_types_typeddict.py  │  test_recursive_generic_typeddict_in_function_1                       │  612            │  632         │  AssertionError   │
│  tests/test_types_typeddict.py  │  test_recursive_generic_typeddict_in_function_2                       │  648            │  663         │  Failed           │
│  tests/test_types_typeddict.py  │  test_recursive_generic_typeddict_in_function_3                       │  681            │  699         │  AssertionError   │
│  tests/test_typing.py           │  test_ns_resolver_push                                                │  188            │  191         │  AssertionError   │
└─────────────────────────────────┴───────────────────────────────────────────────────────────────────────┴─────────────────┴──────────────┴───────────────────┘

```

Reply:
> Seems much more intuitive for later ns entries to take priority. Wondering what the use case is here.

Yeah agreed, it seems pretty weird why it needs to be like this 😅

Reply:
> These fail if we instead do self._namespaces.append(type_ns) + self._namespaces.pop() (which is different from original as the priority is other way around)

Hmm, interesting.

For the dataclass annotations tests, I think we can fix this by reversing the order in which we add the dc bases to the types namespace stack (it's sort of hacky already).

I'm not sure without doing some debugging for the other tests, but I'd certainly be interested in using `self._namespaces.append(type_ns)` and trying to fix the breaking tests with that logic in place.